### PR TITLE
Fix invalid loop range in SymmetricTensor class constructor; add test

### LINKED ISSUE #3674: Possible error in SymmetricTensor constructor
I think that there may be a bug with constructor for the `SymmetricTensor` class that takes a `SymmetricTensor` with a different number type as an argument.

If I'm not mistaken, [this line](https://github.com/dealii/dealii/blob/master/include/deal.II/base/symmetric_tensor.h#L989) should be changed from
```
for (unsigned int i=0; i<n_independent_components; ++i)
```
to
```
for (unsigned int i=0; i<base_tensor_type::dimension; ++i)
```

If I understand correctly then `n_independent_components` corresponds to all of the entries in the `data` `Tensor`, but `base_tensor_type::dimension` corresponds to the number of rows in the underlying `Tensor`.

I someone is willing to confirm this then I'll issue a PR for the change.

### PR DESCRIPTION
**Pull request by jppelteret**
Fixes #3674

### CODE REVIEW

**Comment by jppelteret** [comment]
Confirmation of the bug: Without this small patch to `SymmetricTensor`, this test fails with
```
485: DEAL::Rank 2, Dim 2:  From f  To f ... OK
485: DEAL::Rank 2, Dim 2:  From d  To f ... OK
485: DEAL::Rank 2, Dim 2:  From f  To d ... OK
485: DEAL::Rank 2, Dim 2:  From d  To d ... OK
485: DEAL::Rank 2, Dim 2:  From f  To f ... OK
485: DEAL::Rank 2, Dim 2:  From d  To f ... OK
485: DEAL::Rank 2, Dim 2:  From f  To d ... OK
485: DEAL::Rank 2, Dim 2:  From d  To d ... OK
485: DEAL::Rank 4, Dim 2:  From f  To f ... OK
485: DEAL::Rank 4, Dim 2:  From d  To f ... OK
485: DEAL::Rank 4, Dim 2:  From f  To d ... OK
485: DEAL::Rank 4, Dim 2:  From d  To d ... OK
485: DEAL::Rank 4, Dim 2:  From f  To f ... OK
485: DEAL::Rank 4, Dim 2:  From d  To f ...
485: base/symmetric_tensor_36.debug: RUN failed. ------ Additional output on stdout/stderr:
485:
485:
485: --------------------------------------------------------
485: An error occurred in line <850> of file </Users/jp/Programs/GitHub/deal.II/dealii/include/deal.II/base/tensor.h> in function
485:     ArrayElementType &dealii::internal::TensorSubscriptor::subscript(ArrayElementType *, const unsigned int, dealii::internal::int2type<dim>) [ArrayElementType = dealii::Tensor<1, 3, float>, dim = 3]
485: The violated condition was:
485:     i<dim
485: Additional information:
485:     Index 3 is not in the half-open range [0,3).
485: --------------------------------------------------------
```

File: tests/base/symmetric_tensor_36.cc [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+template<int rank, int dim,
+         template<int,int,typename> class TensorType,
+         typename NumberType1>
+void
+test_fixed_NT_2()
+{
+  test_tensor_constructor<rank,dim,TensorType,NumberType1,float>();
+  test_tensor_constructor<rank,dim,TensorType,NumberType1,double>();
+  // test_tensor_constructor<rank,dim,TensorType,NumberType1,std::complex<double> >();
```
Comment (davydden) [inline tests/base/symmetric_tensor_36.cc]: perhaps you forgot to uncomment this one.

Comment (jppelteret) [inline tests/base/symmetric_tensor_36.cc]: Nope, I did that on purpose. Its a reminder (for me) that one can't convert from a complex<something> to a float/double.

Comment (davydden) [inline tests/base/symmetric_tensor_36.cc]: then please add a comment above

**Comment by jppelteret** [comment]
@davydden I think you'll find the update to your satisfaction.

**Comment by bangerth** [comment]
Looks good. Please rebase to master, add an entry to the changelog in the new format, and merge yourself.

**Comment by davydden** [comment]
first PR with the new `changes` log 🎉

**Pull request merged**